File: 6-IPC-pipe.py
import os
import multiprocessing

# Define the child_process function outside of the parent_child_communication function
def child_process(parent_to_child, child_to_parent):
    # Read message from the parent
    parent_message = parent_to_child.recv()
    print("Child: Received message from parent:", parent_message)

    # Message from child to parent
    child_message = "Hello from child!"
    child_to_parent.send(child_message)
    print("Child: Sent response to parent.")

def parent_child_communication():
    # Create two pipes: one for parent-to-child and one for child-to-parent communication
    parent_to_child, child_to_parent = multiprocessing.Pipe()

    # Message from parent to child
    parent_message = "Hello from parent!"
    parent_to_child.send(parent_message)
    print("Parent: Sent message to child.")

    # Start child process
    p = multiprocessing.Process(target=child_process, args=(parent_to_child, child_to_parent))
    p.start()

    # Read response from the child
    child_response = child_to_parent.recv()
    print("Parent: Received response from child:", child_response)

    # Wait for child to finish
    p.join()

if __name__ == "__main__":
    parent_child_communication()

File: 6-IPC-pipe.c
// #include <stdio.h>
// #include <stdlib.h>
// #include <unistd.h>
// #include <string.h>

// int main() {
//     int pipe1[2], pipe2[2];
//     pid_t pid;
//     char parentMessage[] = "Hello from parent!";
//     char childMessage[] = "Hello from child!";
//     char buffer[100];

//     // Create the pipes
//     if (pipe(pipe1) == -1 || pipe(pipe2) == -1) {
//         perror("Pipe creation failed");
//         exit(1);
//     }

//     // Fork to create child process
//     pid = fork();

//     if (pid < 0) {
//         perror("Fork failed");
//         exit(1);
//     }

//     if (pid > 0) {
//         // Parent process
//         close(pipe1[0]); // Close reading end of pipe1
//         close(pipe2[1]); // Close writing end of pipe2

//         // Send message to child
//         write(pipe1[1], parentMessage, strlen(parentMessage) + 1);
//         close(pipe1[1]); // Close writing end after sending

//         // Read child's response
//         read(pipe2[0], buffer, sizeof(buffer));
//         printf("Parent received: %s\n", buffer);
//         close(pipe2[0]); // Close reading end
//     } else {
//         // Child process
//         close(pipe1[1]); // Close writing end of pipe1
//         close(pipe2[0]); // Close reading end of pipe2

//         // Read message from parent
//         read(pipe1[0], buffer, sizeof(buffer));
//         printf("Child received: %s\n", buffer);
//         close(pipe1[0]); // Close reading end

//         // Send response to parent
//         write(pipe2[1], childMessage, strlen(childMessage) + 1);
//         close(pipe2[1]); // Close writing end
//     }

//     return 0;
// }
